ldb_rotating_reader.go

```diff
@@ -0,0 +1,140 @@
+package ctlstore
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/segmentio/ctlstore/pkg/ldb"
+	"github.com/segmentio/stats/v4"
+	"sync/atomic"
+	"time"
+)
+
+// LDBRotatingReader reads data from multiple LDBs on a rotating schedule.
+// The main benefit is relieving read pressure on a particular ldb file when it becomes inactive,
+// allowing sqlite maintenance
+type LDBRotatingReader struct {
+	active         int32
+	dbs            []RowRetriever
+	schedule       []int8
+	now            func() time.Time
+	tickerInterval time.Duration
+}
+
+type RotationFrequency int
+
+const (
+	// Every30 rotate on 30 minute mark in an hour
+	Every30 RotationFrequency = 30
+	// Every20 rotate on 20 minute marks in an hour
+	Every20 RotationFrequency = 20
+	// Every15 rotate on 15 minute marks in an hour
+	Every15 RotationFrequency = 15
+	// Every10 rotate on 10 minute marks in an hour
+	Every10 RotationFrequency = 10
+	// Every6 rotate on 6 minute marks in an hour
+	Every6 RotationFrequency = 6
+
+	// for simpler migration, the first ldb retains the original name
+	defaultPath = DefaultCtlstorePath + ldb.DefaultLDBFilename
+	ldbFormat   = DefaultCtlstorePath + "ldb-%d.db"
+)
+
+func defaultPaths(count int) []string {
+	paths := []string{defaultPath}
+	for i := 1; i < count; i++ {
+		paths = append(paths, fmt.Sprintf(ldbFormat, i+1))
+	}
+	return paths
+}
+
+// RotatingReader creates a new reader that rotates which ldb it reads from on a rotation frequency with the default location in /var/spool/ctlstore
+func RotatingReader(ctx context.Context, rotationsPerHour RotationFrequency, ldbsCount int) (RowRetriever, error) {
+	return CustomerRotatingReader(ctx, rotationsPerHour, defaultPaths(ldbsCount)...)
+}
+
+// CustomerRotatingReader creates a new reader that rotates which ldb it reads from on a rotation frequency
+func CustomerRotatingReader(ctx context.Context, rotationsPerHour RotationFrequency, ldbPaths ...string) (RowRetriever, error) {
+	r, err := rotatingReader(rotationsPerHour, ldbPaths...)
+	if err != nil {
+		return nil, err
+	}
+	r.setActive()
+	go r.rotate(ctx)
+	return r, nil
+}
+
+func rotatingReader(rotationsPerHour RotationFrequency, ldbPaths ...string) (*LDBRotatingReader, error) {
```

Comment:
This feels a bit odd to me, my interpretation is that rotationsPerHour * RotationFrequency = 60, e.g., if rotation frequency is 30 min, there will be 2 rotations within an hour

Reply:
I think it's a bit of a naming intermixing issue.  `RotationFrequency` is not actually the number of rotations per hour.  I often intermix the physics terms frequency and period, wondering if we have a bit of that going on here.

Frequency is a rate (events / time).  Here we are defining "RotationFrequency" of 6 to mean "rotate every 6 minutes", which I *think* is the period instead of frequency?  My brain is hurting.

Reply:
the variable is named wrong, it's leftover from a previous approach. I'll rename it.

It should be `rotationsEveryXMinutes` or something like that.